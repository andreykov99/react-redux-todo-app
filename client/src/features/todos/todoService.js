import axios from 'axios';

const API_URL = '/api/todos/';

const authHeader = () => {
  const user = JSON.parse(localStorage.getItem('user'));

  if (user && user.token) {
    // For Spring Boot back-end
    return { Authorization: 'Bearer ' + user.token };

    // for Node.js Express back-end
    // return { 'x-access-token': user.accessToken };
  } else {
    return {};
  }
};

//Get todos
const getTodos = async () => {
  const response = await axios.get(API_URL, { headers: authHeader() });
  return response.data;
};

// Add new todo
const addTodo = async ({ title }) => {
  const response = await axios.post(
    API_URL,
    { title },
    { headers: authHeader() }
  );
  return response.data;
};

// Toggle todo completed
const toggleComplete = async ({ id, completed }) => {
  const response = await axios.patch(
    `${API_URL}${id}`,
    { completed },
    {
      headers: authHeader(),
    }
  );
  return response.data;
};

// Delete todo
const deleteTodo = async ({ id }) => {
  const response = await axios.delete(`${API_URL}${id}`, {
    headers: authHeader(),
  });
  return response.data;
};

const todoService = {
  getTodos,
  addTodo,
  toggleComplete,
  deleteTodo,
};

export default todoService;
