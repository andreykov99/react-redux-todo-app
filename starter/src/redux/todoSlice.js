import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';

const api = 'http://localhost:7000/todos';

export const getTodosAsync = createAsyncThunk(
  'todos/getTodosAsync',
  async () => {
    const responce = await fetch(api);
    if (responce.ok) {
      const todos = await responce.json();
      console.log(`#####: todos = ${todos} `);
      return { todos };
    }
  }
);

export const addTodoAsync = createAsyncThunk(
  'todos/addTodoAsync',
  async (payload) => {
    const responce = await fetch(api, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        title: payload.title,
      }),
    });
    if (responce.ok) {
      const todo = await responce.json();
      return { todo };
    }
  }
);

export const toggleCompleteAsync = createAsyncThunk(
  'todos/toggleCompleteAsync',
  async (payload) => {
    const responce = await fetch(`${api}/${payload.id}`, {
      method: 'PATCH',
      headers: { 'Content-Type': 'application/json' },
      body: JSON.stringify({
        completed: payload.completed,
      }),
    });
    if (responce.ok) {
      const todo = await responce.json();
      return { id: todo.id, completed: todo.completed };
    }
  }
);

export const deleteTodoAsync = createAsyncThunk(
  'todos/deleteTodoAsync',
  async (payload) => {
    const responce = await fetch(`${api}/${payload.id}`, {
      method: 'DELETE',
    });
    if (responce.ok) {
      const todos = await responce.json();
      return { todos };
    }
  }
);

export const todoSlice = createSlice({
  name: 'todos',
  initialState: {
    status: 'idle',
    entries: [],
  },
  reducers: {},
  extraReducers: {
    [getTodosAsync.pending]: (state, action) => {
      state.status = 'pending';
    },
    [getTodosAsync.fulfilled]: (state, action) => {
      state.status = 'resolved';
      state.entries = action.payload.todos;
    },
    [addTodoAsync.pending]: (state, action) => {
      state.status = 'pending';
    },
    [addTodoAsync.fulfilled]: (state, action) => {
      state.entries.push(action.payload.todo);
      state.status = 'resolved';
    },
    [toggleCompleteAsync.pending]: (state, action) => {
      state.status = 'pending';
    },
    [toggleCompleteAsync.fulfilled]: (state, action) => {
      const index = state.entries.findIndex(
        (todo) => todo.id === action.payload.id
      );
      state.entries[index].completed = action.payload.completed;
      state.status = 'resolved';
    },
    [deleteTodoAsync.pending]: (state, action) => {
      state.status = 'pending';
    },
    [deleteTodoAsync.fulfilled]: (state, action) => {
      state.status = 'resolved';
      state.entries = action.payload.todos;
    },
  },
});

export const { addTodo, toggleComplete, deleteTodo } = todoSlice.actions;

export default todoSlice.reducer;
