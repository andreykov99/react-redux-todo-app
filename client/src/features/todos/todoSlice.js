import { createSlice, createAsyncThunk } from '@reduxjs/toolkit';
import todoService from './todoService';

export const getTodosAsync = createAsyncThunk(
  'todos/getTodosAsync',
  async (thunkAPI) => {
    try {
      return todoService.getTodos();
    } catch (error) {
      const message =
        (error.response &&
          error.response.data &&
          error.response.data.message) ||
        error.message ||
        error.toString();
      return thunkAPI.rejectWithValue(message);
    }
  }
);

export const addTodoAsync = createAsyncThunk(
  'todos/addTodoAsync',
  async (todo, thunkAPI) => {
    try {
      return todoService.addTodo(todo);
    } catch (error) {
      const message =
        (error.response &&
          error.response.data &&
          error.response.data.message) ||
        error.message ||
        error.toString();
      return thunkAPI.rejectWithValue(message);
    }
  }
);

export const toggleCompleteAsync = createAsyncThunk(
  'todos/toggleCompleteAsync',
  async (todo, thunkAPI) => {
    try {
      return todoService.toggleComplete(todo);
    } catch (error) {
      const message =
        (error.response &&
          error.response.data &&
          error.response.data.message) ||
        error.message ||
        error.toString();
      return thunkAPI.rejectWithValue(message);
    }
  }
);

export const deleteTodoAsync = createAsyncThunk(
  'todos/deleteTodoAsync',
  async (todo, thunkAPI) => {
    try {
      return todoService.deleteTodo(todo);
    } catch (error) {
      const message =
        (error.response &&
          error.response.data &&
          error.response.data.message) ||
        error.message ||
        error.toString();
      return thunkAPI.rejectWithValue(message);
    }
  }
);

export const todoSlice = createSlice({
  name: 'todos',
  initialState: {
    status: 'idle',
    entries: [],
  },
  reducers: {},
  extraReducers: {
    [getTodosAsync.pending]: (state, action) => {
      state.status = 'pending';
    },
    [getTodosAsync.fulfilled]: (state, action) => {
      state.status = 'resolved';
      state.entries = action.payload;
    },
    [addTodoAsync.pending]: (state, action) => {
      state.status = 'pending';
    },
    [addTodoAsync.fulfilled]: (state, action) => {
      state.entries.push(action.payload);
      state.status = 'resolved';
    },
    [toggleCompleteAsync.pending]: (state, action) => {
      state.status = 'pending';
    },
    [toggleCompleteAsync.fulfilled]: (state, action) => {
      const index = state.entries.findIndex(
        (todo) => todo._id === action.payload._id
      );
      state.entries[index].completed = action.payload.completed;
      state.status = 'resolved';
    },
    [deleteTodoAsync.pending]: (state, action) => {
      state.status = 'pending';
    },
    [deleteTodoAsync.fulfilled]: (state, action) => {
      state.status = 'resolved';
      state.entries = state.entries.filter(
        (todo) => todo._id !== action.payload.id
      );
    },
  },
});

export const { addTodo, toggleComplete, deleteTodo } = todoSlice.actions;

export default todoSlice.reducer;
