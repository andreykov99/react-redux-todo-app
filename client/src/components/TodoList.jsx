import React, { useEffect } from 'react';
import TodoItem from './TodoItem';
import { useSelector, useDispatch } from 'react-redux';
import { getTodosAsync } from '../features/todos/todoSlice';

const TodoList = () => {
  const dispatch = useDispatch();
  const todos = useSelector((state) => state.todos.entries);

  useEffect(() => {
    dispatch(getTodosAsync());
  }, [dispatch]);

  return (
    <ul className="list-group">
      {todos.map(({ _id: id, title, completed }) => (
        <TodoItem key={id} id={id} title={title} completed={completed} />
      ))}
    </ul>
  );
};

export default TodoList;
