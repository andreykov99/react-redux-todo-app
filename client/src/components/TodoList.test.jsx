import { render } from '@testing-library/react';
import TodoList from './TodoList';
import * as reduxHooks from 'react-redux';
import * as actions from '../features/todos/todoSlice';

jest.mock('react-redux');
const mockedSelector = jest.spyOn(reduxHooks, 'useSelector');
const mockedDispatch = jest.spyOn(reduxHooks, 'useDispatch');
describe('TodoList component', () => {
  it('should be rendered with todos', () => {
    mockedSelector.mockReturnValue([
      { id: '123', title: 'Todo1', completed: false },
    ]);
    const dispatch = jest.fn();
    mockedDispatch.mockReturnValue(dispatch);
    const mockGetTodos = jest.spyOn(actions, 'getTodosAsync');
    const view = render(<TodoList />);
    expect(view).toMatchSnapshot();
    expect(mockGetTodos).toHaveBeenCalled();
  });
  it('should be rendered without todos', () => {
    mockedSelector.mockReturnValue([]);
    const dispatch = jest.fn();
    mockedDispatch.mockReturnValue(dispatch);
    const mockGetTodos = jest.spyOn(actions, 'getTodosAsync');
    const view = render(<TodoList />);
    expect(view).toMatchSnapshot();
    expect(mockGetTodos).toHaveBeenCalled();
  });
});
