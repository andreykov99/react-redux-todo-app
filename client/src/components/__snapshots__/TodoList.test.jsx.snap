// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TodoList component should be rendered with todos 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="list-group"
      >
        <li
          class="list-group-item false"
        >
          <div
            class="d-flex justify-content-between"
          >
            <span
              class="d-flex align-items-center"
            >
              <input
                class="mr-3"
                type="checkbox"
              />
              Todo1
            </span>
            <button
              class="btn btn-danger"
            >
              Delete
            </button>
          </div>
        </li>
      </ul>
    </div>
  </body>,
  "container": <div>
    <ul
      class="list-group"
    >
      <li
        class="list-group-item false"
      >
        <div
          class="d-flex justify-content-between"
        >
          <span
            class="d-flex align-items-center"
          >
            <input
              class="mr-3"
              type="checkbox"
            />
            Todo1
          </span>
          <button
            class="btn btn-danger"
          >
            Delete
          </button>
        </div>
      </li>
    </ul>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`TodoList component should be rendered without todos 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <ul
        class="list-group"
      />
    </div>
  </body>,
  "container": <div>
    <ul
      class="list-group"
    />
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
